test/hotspot/jtreg/compiler/lib/template_framework/Template.java

```diff
@@ -608,14 +607,184 @@ static <T1, T2, T3> Template.ThreeArgs<T1, T2, T3> make(String arg1Name, String
+     *     ),
+     *     // CODE3: more code in the outer scope, names and hashtags from CODE2 are
+     *     //        not available any more because of the non-transparent "scope".
+     *     transparentScope(
+     *       // CODE4: some code in the inner "transparentScope". Names, hashtags and setFuelCost
+     *       //        escape the "transparentScope" and are still available after the "transparentScope"
+     *       //        closes.
+     *     )
+     *     // CODE5: we still have access to names and hashtags from CODE4.
+     * ));
+     * }
+     *
+     * @param tokens A list of tokens, which can be {@link String}s, boxed primitive types
+     *               (for example {@link Integer}), any {@link Token}, or {@link List}s
+     *               of any of these.
+     * @return The {@link ScopeToken} which captures the list of validated {@link Token}s.
+     * @throws IllegalArgumentException if the list of tokens contains an unexpected object.
+     */
+    static ScopeToken scope(Object... tokens) {
+        return new ScopeTokenImpl(TokenParser.parse(tokens), true, true, true);
+    }
+
+    /**
+     * Creates a {@link ScopeToken} which represents a scope that is transparent for
+     * {@link DataName}s and {@link StructuralName}s, as well as hashtag-replacements and
+     * {@link setFuelCost}. This means that any such name, hashtag-replacement or
+     * {@link setFuelCost} declared inside the scope escapes that scope and is still
+     * available in the outer scope.
+     * The scope is formed from a list of tokens, which can be {@link String}s,
+     * boxed primitive types (for example {@link Integer} or auto-boxed {@code int}),
+     * any {@link Token}, or {@link List}s of any of these.
+     *
+     * <p>
+     * If you require a scope that is non-transparent for some or all of the above, consider
+     * using {@link scope}, {@link nameScope}, {@link hashtagScope}, or {@link setFuelCostScope}.
+     *
      * @param tokens A list of tokens, which can be {@link String}s, boxed primitive types
      *               (for example {@link Integer}), any {@link Token}, or {@link List}s
      *               of any of these.
-     * @return The {@link TemplateBody} which captures the list of validated {@link Token}s.
+     * @return The {@link ScopeToken} which captures the list of validated {@link Token}s.
      * @throws IllegalArgumentException if the list of tokens contains an unexpected object.
      */
-    static TemplateBody body(Object... tokens) {
-        return new TemplateBody(TokenParser.parse(tokens));
+    static ScopeToken transparentScope(Object... tokens) {
+        return new ScopeTokenImpl(TokenParser.parse(tokens), false, false, false);
+    }
+
+    /**
+     * Creates a {@link ScopeToken} which represents a scope that is non-transparent for
+     * {@link DataName}s and {@link StructuralName}s, but transparent for hashtag-replacements
+     * and {@link setFuelCost}.
```

Comment:
Adapting from above suggestions:

```suggestion
     * Creates a {@link ScopeToken} that represents a scope that is non-transparent for
     * hashtag-replacements (i.e. cannot escape), but transparent for {@link DataName}s 
     * and {@link StructuralName}s and {@link #setFuelCost} (i.e. available in outer scope).
     * 
     * <p>
```